tests/Pyxsim/pyxsim.py

```diff
@@ -0,0 +1,466 @@
+        (tile, p, bit, mask) = parse_port(port)
+        val = self._xsi.sample_port_pins(tile, p, mask)
+        if bit:
+            val >>= bit
+        return val
+
+    # TODO make this pin*s*
+    def drive_periph_pin(self, pin, val):
+        (periph, pin, mask) = parse_periph_pin(pin)
+        self._xsi.drive_periph_pin(periph, pin, mask, val)
+
+    # TODO make this pin*s*
+    def sample_periph_pin(self, pin):
+        (periph, pin, mask) = parse_periph_pin(pin)
+        return self._xsi.sample_periph_pin(periph, pin, mask)
+
+    def terminate(self):
+        self.terminate_flag = True
+        self._wait(lambda x: False)
+
+    def run(self):
+        args = self._args
+        try:
+            self._fn(*args)
+        except:
+            self.had_exception = True
+            sys.stderr.write("---------Exception in simthread--------------\n")
+            traceback.print_exc()
+            sys.stderr.write("---------------------------------------------\n")
+        self.resume_condition = lambda x: False
+        self.complete_event.set()
+
+
+class Xsi(object):
+    def __init__(self, xe_path=None, simargs=[], appargs=[]):
+        self.xsim = c_void_p()
+        self.xe_path = xe_path
+        args = " ".join(
+            ['"{}"'.format(x) for x in (simargs + [self.xe_path] + appargs)]
+        )
+        if platform_is_windows:
+            args = args.replace("\\", "/")
+        c_args = c_char_p(args.encode("utf-8"))
+        status = xsi_lib.xsi_create(byref(self.xsim), c_args)
+        self._plugins = []
+        self._simthreads = []
+        self._time = 0
+
+    def register_plugin(self, plugin):
+        self._plugins.append(plugin)
+
+    def register_simthread(self, fn):
+        if isinstance(fn, tuple):
+            xs = list(fn)
+            fn = xs[0]
+            args = xs[1:]
+        else:
+            args = []
+
+        simthread = SimThreadImpl(self, fn, args)
+        simthread.daemon = True
+        self._simthreads.append(simthread)
+
+    def wait_for_simthread(self, simthread):
+        simthread.complete_event.wait()
+        simthread.complete_event.clear()
+        if simthread.had_exception:
+            raise TestError("Simthread encoutered an exception")
+
+    def clock(self):
+        status = xsi_lib.xsi_clock(self.xsim)
+        self._time += self._time_step
+        if XsiStatus.is_valid(status):
+            for plugin in self._plugins:
+                plugin.clock(self)
+            for simthread in self._simthreads:
+                if simthread.resume_condition(self):
+                    simthread.resume_event.set()
+                    self.wait_for_simthread(simthread)
+                    if simthread.terminate_flag:
+                        return XsiStatus.DONE
+
+        return status
+
+    def get_time(self):
+        return self._time
+
+    def run(self):
+        self.xe = Xe(self.xe_path)
+        self._time_step = float(pow(10, 3)) / self.xe.freq
```

Comment:
Does this really need to be ```float(pow(10, 3))```?? Surely ```1000.0``` would be as clear, if not clearer? Jus' sayin'...

Reply:
Agreed. Not sure where this came from...